wire-ios-sync-engine/Source/UserSession/URLActionProcessors/DeepLinkURLActionProcessor.swift

```diff
@@ -36,96 +36,129 @@ class DeepLinkURLActionProcessor: URLActionProcessor {
-                            ) { [weak self] response in
-
-                                guard let strongSelf = self else { return }
-
-                                switch response {
-                                case .success(let conversation):
-                                    strongSelf.synchronise(conversation) { result in
-                                        DispatchQueue.main.async {
-                                            switch result {
-                                            case .success(let syncConversation):
-                                                delegate.showConversation(syncConversation, at: nil)
-                                            case .failure(let error):
-                                                delegate.failedToPerformAction(urlAction, error: error)
-                                            }
-                                        }
-                                    }
-                                case .failure(let error):
-                                    delegate.failedToPerformAction(urlAction, error: error)
-                                }
-
-                                delegate.completedURLAction(urlAction)
-                            }
-                        }
-                    }
+        case .openConversation(let id):
+            handleOpenConversation(id: id, delegate: delegate)
+
+        case .openUserProfile(let id):
+            handleOpenUserProfile(id: id, delegate: delegate)
+
+        default:
+            delegate?.completedURLAction(urlAction)
+        }
+    }
+
+    private func handleJoinConversation(key: String, code: String, urlAction: URLAction, delegate: PresentationDelegate?) {
+        ZMConversation.fetchIdAndName(
+            key: key,
+            code: code,
+            transportSession: transportSession,
+            contextProvider: contextProvider
+        ) { [weak self] response in
+            guard let self, let delegate else {
+                return
+            }
+
+            let viewContext = self.contextProvider.viewContext
 
-                case .failure(let error):
-                    delegate.failedToPerformAction(urlAction, error: error)
+            switch response {
+            case .success((let conversationId, let conversationName, let hasPassword)):
+                if let conversation = ZMConversation.fetch(with: conversationId, in: viewContext),
+                   conversation.isSelfAnActiveMember {
+                    delegate.showConversation(conversation, at: nil)
                     delegate.completedURLAction(urlAction)
+                } else if hasPassword {
+                    self.handlePasswordPrompt(for: conversationName, key: key, code: code, delegate: delegate)
+                } else {
+                    self.handleJoinWithoutPassword(for: conversationName, key: key, code: code, urlAction: urlAction, delegate: delegate)
                 }
```

Comment:
Could this be extracted to new func?

Reply:
This is a bit tricky to pull off so I can skip for now.